### ISSUE TITLE
Type param + Union + Default = dispatch error

### ISSUE THREAD

**Issue by alyst** [JuliaLang/julia#15517]
This code:
```julia
type A end
Base.convert{F}(::Type{A}, x::F, v::Union{F,Vector{F}} = one(F)) = A()
```
on Julia v0.4.3 gives
```
julia> convert(A, 0.1)
A()

julia> convert(A, 0.1, 0.2)
A()

julia> convert(A, 0.1, [0.2])
ERROR: MethodError: `convert` has no method matching convert(::Type{A}, ::Float64, ::Array{Float64,1})
This may have arisen from a call to the constructor A(...),
since type constructors fall back to convert methods.
Closest candidates are:
  convert{F}(::Type{A}, ::F)
  convert{F}(::Type{A}, ::F, ::Union{Array{F,1},F})
  call{T}(::Type{T}, ::Any)
  ...
```
I would expect that the last call also works. If I define `convert()`, where `F` is replaced by `Float64`, it works.

On v0.5 from the juliabox.org
```
julia> convert(A, 0.1)                                                                                                                                                                                                                                                                                                    
A(Error showing value of type A:                                                                                                                                                                                                                                                                                          
ERROR: MethodError: `one` has no method matching one(::Type{A})                                                                                                                                                                                                                                                           
 [inlined code] from ./none:1                                                                                                                                                                                                                                                                                             
 in Type(Type{Pair{A,B}}, Symbol, A) at ./operators.jl:507                                                                                                                                                                                                                                                                
 in show_default(Base.#show_default, IOContext{Base.Terminals.TTYTerminal}, Any) at ./show.jl:80                                                                                                                                                                                                                          
 [inlined code] from ./show.jl:73                                                                                                                                                                                                                                                                                         
 in showcompact(Base.#showcompact, Base.Terminals.TTYTerminal, A) at ./show.jl:1432                                                                                                                                                                                                                                       
 in writemime(Base.Multimedia.#writemime, Base.Terminals.TTYTerminal, MIME{symbol("text/plain")}, A) at ./replutil.jl:4                                                                                                                                                                                                   
 [inlined code] from ./expr.jl:8                                                                                                                                                                                                                                                                                          
 in display(Base.Multimedia.#display, Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, MIME{symbol("text/plain")}, Any) at ./REPL.jl:114                                                                                                                                                                                    
 [inlined code] from ./expr.jl:8                                                                                                                                                                                                                                                                                          
 in display(Base.Multimedia.#display, Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, Any) at ./REPL.jl:117                                                                                                                                                                                                                
 [inlined code] from ./multimedia.jl:151                                                                                                                                                                                                                                                                                  
 in display(Base.Multimedia.#display, Any) at ./multimedia.jl:163                                                                                                                                                                                                                                                         
 in print_response(Base.REPL.#print_response, Base.Terminals.TTYTerminal, Any, Void, Bool, Bool, Void) at ./REPL.jl:134                                                                                                                                                                                                   
 in print_response(Base.REPL.#print_response, Base.REPL.LineEditREPL, Any, Void, Bool, Bool) at ./REPL.jl:121                                                                                                                                                                                                             
 in #18(Base.REPL.##18#19{Bool,Base.#parse_input_line,Base.REPL.LineEditREPL,Base.LineEdit.Prompt}, Base.LineEdit.MIState, Base.AbstractIOBuffer{Array{UInt8,1}}, Bool) at ./REPL.jl:626                                                                                                                                  
 in run_interface at ./LineEdit.jl:1611                                                                                                                                                                                                                                                                                   
 in run_frontend at ./REPL.jl:859                                                                                                                                                                                                                                                                                         
 in run_repl(Base.REPL.#run_repl, Base.REPL.LineEditREPL, Base.##479#480) at ./REPL.jl:167                                                                                                                                                                                                                                
 in _start at ./client.jl:342                                                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                          
julia> convert(A, 0.1, 0.1)                                                                                                                                                                                                                                                                                               
A(Error showing value of type A:                                                                                                                                                                                                                                                                                          
ERROR: MethodError: `one` has no method matching one(::Type{A})                                                                                                                                                                                                                                                           
 [inlined code] from ./none:1                                                                                                                                                                                                                                                                                             
 in Type(Type{Pair{A,B}}, Symbol, A) at ./operators.jl:507                                                                                                                                                                                                                                                                
 in show_default(Base.#show_default, IOContext{Base.Terminals.TTYTerminal}, Any) at ./show.jl:80                                                                                                                                                                                                                          
 [inlined code] from ./show.jl:73                                                                                                                                                                                                                                                                                         
 in showcompact(Base.#showcompact, Base.Terminals.TTYTerminal, A) at ./show.jl:1432                                                                                                                                                                                                                                       
 in writemime(Base.Multimedia.#writemime, Base.Terminals.TTYTerminal, MIME{symbol("text/plain")}, A) at ./replutil.jl:4                                                                                                                                                                                                   
 [inlined code] from ./expr.jl:8                                                                                                                                                                                                                                                                                          
 in display(Base.Multimedia.#display, Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, MIME{symbol("text/plain")}, Any) at ./REPL.jl:114                                                                                                                                                                                    
 [inlined code] from ./expr.jl:8                                                                                                                                                                                                                                                                                          
 in display(Base.Multimedia.#display, Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, Any) at ./REPL.jl:117                                                                                                                                                                                                                
 [inlined code] from ./multimedia.jl:151                                                                                                                                                                                                                                                                                  
 in display(Base.Multimedia.#display, Any) at ./multimedia.jl:163                                                                                                                                                                                                                                                         
 in print_response(Base.REPL.#print_response, Base.Terminals.TTYTerminal, Any, Void, Bool, Bool, Void) at ./REPL.jl:134                                                                                                                                                                                                   
 in print_response(Base.REPL.#print_response, Base.REPL.LineEditREPL, Any, Void, Bool, Bool) at ./REPL.jl:121                                                                                                                                                                                                             
 in #18(Base.REPL.##18#19{Bool,Base.#parse_input_line,Base.REPL.LineEditREPL,Base.LineEdit.Prompt}, Base.LineEdit.MIState, Base.AbstractIOBuffer{Array{UInt8,1}}, Bool) at ./REPL.jl:626                                                                                                                                  
 in run_interface at ./LineEdit.jl:1611                                                                                                                                                                                                                                                                                   
 in run_frontend at ./REPL.jl:859                                                                                                                                                                                                                                                                                         
 in run_repl(Base.REPL.#run_repl, Base.REPL.LineEditREPL, Base.##479#480) at ./REPL.jl:167                                                                                                                                                                                                                                
 in _start at ./client.jl:342                                                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                          
julia> convert(A, 0.1, [0.1])                                                                                                                                                                                                                                                                                             
ERROR: MethodError: `convert` has no method matching convert(::Type{A}, ::Float64, ::Array{Float64,1})                                                                                                                                                                                                                    
This may have arisen from a call to the constructor A(...),                                                                                                                                                                                                                                                               
since type constructors fall back to convert methods.                                                                                                                                                                                                                                                                     
Closest candidates are:                                                                                                                                                                                                                                                                                                   
  convert{F}(::Type{A}, ::F)                                                                                                                                                                                                                                                                                              
  convert{F}(::Type{A}, ::F, ::Union{Array{F,1},F})
  (::Type{BoundsError})(::ANY, ::ANY)
  ...                                                                                                                                                                                                                                                                                                                     
 in eval at ./boot.jl:267                                                                                                                                                                                                                                                                                                 
```

**Comment by mbauman** [comment]
Type parameters currently match greedily.  They work just fine when all the components of the Union are orthogonal, but when multiple components can match, there can be surprises. In your example, `F` could *either* be `Array{Float64}` or `Float64` if the typevar is resolved in the Union before `x::F`.  Somewhat unrelatedly, you're seeing an error in the display code on 0.5 because you are overriding the default no-op convert with your method, but haven't defined `one(::Type{A})`.

This is a duplicate of https://github.com/JuliaLang/julia/issues/3738.

**Comment by alyst** [comment]
Thanks for the detailed explanation. While the more advanced approach are not implemented, is there any way to detect and warn the user about "non-orthogonal" unions? With the current error message it's easy to be confused and waste time searching for the bug in your own declarations (which are not as simple as in the example).

**Comment by mbauman** [comment]
Yes, I agree that would be a nice workaround.  I tried it at one point, but if I remember correctly I got stuck since Julia can create and use non-orthogonal unions in inference without any trouble.  It's just problematic in function definitions… so my general approach wasn't workable.  It should be possible to just check when defining functions.  There's also a documentation issue open about this, I think… and hopefully #8974 will fix everything.
